decorate_test.go

```diff
@@ -0,0 +1,463 @@
+
+		require.NoError(t, c.Decorate(func(a *A) *A { return &A{name: a.name + "'"} }))
+		require.NoError(t, child.Decorate(func(a *A) *A { return &A{name: a.name + "'"} }))
+
+		assert.NoError(t, c.Invoke(func(a *A) {
+			assert.Equal(t, "A'", a.name, "expected decorated name in parent")
+		}))
+
+		assert.NoError(t, child.Invoke(func(a *A) {
+			assert.Equal(t, "A''", a.name, "expected double-decorated name in child")
+		}))
+
+		sibling := c.Scope("sibling")
+		grandchild := child.Scope("grandchild")
+		sibling.Invoke(func(a *A) {
+			assert.Equal(t, "A'", a.name, "expected single-decorated name in sibling")
+		})
+		grandchild.Invoke(func(a *A) {
+			assert.Equal(t, "A''", a.name, "expected double-decorated name in grandchild")
+		})
+	})
+
+	t.Run("decorate with In struct", func(t *testing.T) {
+		t.Parallel()
+
+		type A struct {
+			Name string
+		}
+		type B struct {
+			dig.In
+
+			A *A
+			B string `name:"b"`
+		}
+
+		type C struct {
+			dig.Out
+
+			A *A
+			B string `name:"b"`
+		}
+
+		c := dig.New()
+		require.NoError(t, c.Provide(func() *A { return &A{Name: "A"} }))
+		require.NoError(t, c.Provide(func() string { return "b" }, dig.Name("b")))
+
+		require.NoError(t, c.Decorate(func(b B) C {
+			return C{
+				A: &A{
+					Name: b.A.Name + "'",
+				},
+				B: b.B + "'",
+			}
+		}))
+
+		assert.NoError(t, c.Invoke(func(b B) {
+			assert.Equal(t, "A'", b.A.Name)
+			assert.Equal(t, "b'", b.B)
+		}))
+	})
+
+	t.Run("decorate with value groups", func(t *testing.T) {
+		type Params struct {
+			dig.In
+
+			Animals []string `group:"animals"`
+		}
+
+		type Result struct {
+			dig.Out
+
+			Animals []string `group:"animals"`
+		}
+
+		c := dig.New()
+		require.NoError(t, c.Provide(func() string { return "dog" }, dig.Group("animals")))
+		require.NoError(t, c.Provide(func() string { return "cat" }, dig.Group("animals")))
+		require.NoError(t, c.Provide(func() string { return "alpaca" }, dig.Group("animals")))
+
+		assert.NoError(t, c.Decorate(func(p Params) Result {
+			animals := p.Animals
+			for i := 0; i < len(animals); i++ {
+				animals[i] = "good " + animals[i]
+			}
+			return Result{
+				Animals: animals,
+			}
+		}))
+
+		assert.NoError(t, c.Invoke(func(p Params) {
+			assert.Contains(t, p.Animals, "good dog")
+		}))
+	})
+
+	t.Run("decorate with optional parameter", func(t *testing.T) {
+		c := dig.New()
+
+		type A struct{}
+		type Param struct {
+			dig.In
+
+			Values []string `group:"values"`
+			A      *A       `optional:"true"`
+		}
+
+		type Result struct {
+			dig.Out
+
+			Values []string `group:"values"`
+		}
+
+		require.NoError(t, c.Provide(func() string { return "a" }, dig.Group("values")))
+		require.NoError(t, c.Provide(func() string { return "b" }, dig.Group("values")))
+
+		require.NoError(t, c.Decorate(func(p Param) Result {
+			return Result{
+				Values: append(p.Values, "c"),
+			}
+		}))
+
+		assert.NoError(t, c.Invoke(func(p Param) {
+			assert.Equal(t, 3, len(p.Values))
```

Comment:
For this one, should we check if A is nil, and assign it a value for the invoke?